picard/plugin.py

```diff
@@ -421,24 +415,63 @@ def install_plugin(self, path, overwrite_confirm=None, plugin_name=None,
                         raise
                 elif os.path.isfile(path):
                     dst = os.path.join(USER_PLUGIN_DIR, os.path.basename(path))
-                    if update:
-                        dst += '.update'
-                        if os.path.isfile(dst):
-                            os.remove(dst)
                     shutil.copy2(path, dst)
                 elif os.path.isdir(path):
                     dst = os.path.join(USER_PLUGIN_DIR, plugin_name)
-                    if update:
-                        dst += '.update'
-                        if os.path.isdir(dst):
-                            shutil.rmtree(dst)
                     shutil.copytree(path, dst)
-                if not update:
-                    installed_plugin = self.load_plugin(zip_plugin or plugin_name, USER_PLUGIN_DIR)
-                    if installed_plugin is not None:
-                        self.plugin_installed.emit(installed_plugin, False)
-                else:
-                    self.plugin_updated.emit(plugin_name, False)
+                installed_plugin = self.load_plugin(zip_plugin or plugin_name, USER_PLUGIN_DIR)
+                if installed_plugin is not None:
+                    self.plugin_installed.emit(installed_plugin, False)
+            except (OSError, IOError):
+                log.warning("Unable to copy %s to plugin folder %s" % (path, USER_PLUGIN_DIR))
+
+    def update_plugin(self, path, overwrite_confirm=None, plugin_name=None,
+                       plugin_data=None):
+
+        zip_plugin = False
```

Comment:
Most of the code is duplicated from above... that's bad.
Instead create a function with all the common code, called by those two.